plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftHiveMetastore.java

```diff
@@ -2040,22 +2040,37 @@ private ThriftMetastoreClient createMetastoreClient(HiveIdentity identity)
 
         String username = identity.getUsername().orElseThrow(() -> new IllegalStateException("End-user name should exist when metastore impersonation is enabled"));
         if (authenticationEnabled) {
-            String delegationToken;
             try {
-                delegationToken = delegationTokenCache.getUnchecked(username);
+                return createMetastoreClientWithAuthentication(username, false);
             }
-            catch (UncheckedExecutionException e) {
-                throwIfInstanceOf(e.getCause(), TrinoException.class);
-                throw e;
+            catch (TException e) {
+                log.debug(e, "It may be a delegation token permission failure, so try again with a new delegation token");
+                return createMetastoreClientWithAuthentication(username, true);
             }
-            return clientProvider.createMetastoreClient(Optional.of(delegationToken));
         }
 
         ThriftMetastoreClient client = createMetastoreClient();
         setMetastoreUserOrClose(client, username);
         return client;
     }
 
+    private ThriftMetastoreClient createMetastoreClientWithAuthentication(String username, boolean refresh)
+        throws TException
+    {
+        if (refresh) {
+            delegationTokenCache.refresh(username);
```

Comment:
refresh might be async (under what conditions?). it should be simpler to `invalidate`, but this would be vulnerable to https://github.com/trinodb/trino/issues/10512 / https://github.com/google/guava/issues/1881.

cc @kokosing

Reply:
I think we should not use `refresh`. It makes it hard to reason about the current state. Notice that we don't know if below `getUnchecked` would return new or old value.

Reply:
With the link below, it was confirmed that `refresh` function works as async. Therefore, it has been modified from `refresh` function to `put` function.

* https://github.com/google/guava/blob/master/guava/src/com/google/common/cache/LocalCache.java#L2364